components/coding/profileV2/AchievementComponent.tsx

```diff
@@ -1,172 +1,161 @@
-          codingBadges: unit.codingBadges.map((badge) => {
-            if (badge.id == inputBadge.id) {
-              return {
-                ...badge,
-                isAwarded: !badge.isAwarded,
-              };
-            } else {
-              return badge;
-            }
-          }),
-        };
-      });
 
-      return updatedData;
+  const [saveAddedBadges] = useMutation(INSERT_USER_CODING_BADGES, {
+    refetchQueries: [{ query: FETCH_CODING_BADGES }],
+  });
+  const [saveRemovedBadges] = useMutation(DELETE_USER_CODING_BADGES, {
+    refetchQueries: [{ query: FETCH_CODING_BADGES }],
+  });
+
+  const separateBadges = (
+    changedBadgesList: CodingBadge[],
+    initialBadgeList: CodingBadge[]
+  ) => {
+    const addedBadgesTemp: CodingBadge[] = [];
+    const removedBadgesTemp: CodingBadge[] = [];
+    changedBadgesList.forEach((badge: CodingBadge) => {
+      const index = initialBadgeList.findIndex(
+        (initialBadge) => initialBadge.id === badge.id
+      );
+      if (
+        badge.user_coding_badges.length > 0 &&
+        initialBadgeList.find((initBadge) => badge.id == initBadge.id)
+          .user_coding_badges.length === 0
+      ) {
+        addedBadgesTemp.push(badge);
+      } else if (
+        badge.user_coding_badges.length === 0 &&
+        initialBadgeList[index].user_coding_badges.length > 0
+      ) {
+        removedBadgesTemp.push(badge);
+      }
+    });
+    const addedBadges = addedBadgesTemp.map((badge) => {
+      return { badgeId: badge.id, userId: userId };
     });
+    const removedBadges = removedBadgesTemp.map((badge) => {
+      return { badgeId: badge.id, userId: userId };
+    });
+    return { addedBadges, removedBadges };
   };
 
-  useEffect(() => {
-    if (data) {
-      const updatedData = transformUserBadgeData(data);
+  // this function will take in the original list of badge data (fetched from the query), and the edited badge data
+  // it should output an object containing two arrays: a list of badgeIds to add, and a list of userBadgeIds to remove
+  //  the diff between the two input arrays are then send them to the database to delete and upsert
+  const findBadgeDiff = (
+    initialSet: IntroCourseUnit[],
+    currentSet: IntroCourseUnit[]
+  ) => {
+    const initialBadgeArray = initialSet.flatMap((unit) =>
+      unit.coding_badges.map((badge) => badge)
+    );
+    const finalBadgeArray = currentSet.flatMap((unit) =>
+      unit.coding_badges.map((badge) => badge)
+    );
 
-      setTransformedData(updatedData);
-    }
-  }, [data]);
+    const changedBadgesList = finalBadgeArray.filter((finalBadge) => {
+      const initialBadge = initialBadgeArray.find(
+        (initialBadge) =>
+          finalBadge.id === initialBadge.id &&
+          finalBadge.user_coding_badges === initialBadge.user_coding_badges
+      );
+      return !initialBadge;
+    });
+    //
+
+    return {
+      initialBadgeArray,
+      finalBadgeArray,
+      changedBadgesList,
+    };
```

Comment:
findBadgeDiff takes the difference between the queried badges (initial) and updated badges by the student/coach (finalbadge) and creates a new list containing only the badges that have changes.